### DISCUSSION TITLE
SEB 3.2.0 Win x64 - Virtual Machine Detected / Allow Virtual Machine = True

### DISCUSSION THREAD

**Discussion by UncleRaymondo** [SafeExamBrowser/seb-win-refactoring#103] [General]
Hi,

Trying to run SEB 3.2.0 with Moodle LMS. When students click Launch SEB, it launches then comes back with Virtual Machine Detected. The configuration is configured to Allow Virtual Machines and to Ignore Display Errors. I can see in the XML File that Allow Virtual Machines is also set to True, but the error keeps appearing then quitting.

Looking at some previous issues on here, I can see you requested some commands be run. I have attached the output of those commands to this issue

[pnpEntity.txt](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6775560/pnpEntity.txt)
[NetworkAdapterConfigurationFormatted.txt](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6775561/NetworkAdapterConfigurationFormatted.txt)

This host machine of which where SEB is running from is a VMWare Horizon Client / Virtual Desktop. Students will be using Horizon to get access to their environments of which each will have this host plus numerous virtual machines within the host.

Appreciate your prompt assistance as we have the students starting next week.

Cheers,
Ray

**Comment by dbuechel** [comment]
Are you sure that allow virtual machine is correctly set and that your configuration is in the end actually used by your students? Because the check is pretty failsafe, if the flag is set to allow virtual machines, the entire check is simply skipped, so it looks like your students are still using a configuration where virtual machines are not allowed.

**Comment by UncleRaymondo** [comment]
Hi Damian,

Yep, using the config via a template inside Moodle LMS attached to a Quiz. First thing I checked was that XML to see if that was actually being set to true from the config utility. The config is configured for starting an exam which is applied as a template in Moodle, attached to a Quiz and launches when you select Launch Safe Exam Browser.

I've attached the XML content from the Template in Moodle. I can see in the template the following:

`<key>allowVirtualMachine</key><true/>`

[WSU_SEB_Template.txt](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6780756/WSU_SEB_Template.txt)

Appreciate your help.

Cheers,
Ray

**Comment by dbuechel** [comment]
I think we'll need to check the log file of the runtime to see what's happening in detail. You can find the log files under `%LocalAppData%\SafeExamBrowser\Logs`. In the respective log of the runtime, there should be entries detailing exactly which configuration file is loaded and whether the virtual machine policy is validated or not. If it is validated, then the flag `allowVirtualMachine` is still set to false and we'd need to find out why.

**Comment by UncleRaymondo** [comment]
Hi Damian,

Please see attached...

[2021-07-08_17h21m40s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6782442/2021-07-08_17h21m40s_Runtime.log)

I've made sure there's no stale SEB Config file inside the %AppData%\SafeExamBrowser folder as I noticed in the first attempt logs that it loaded this file (it's not clear in the logs if it's loading the one from Moodle). Deleting these files and performing a test again still reports the same error.

Cheers,
Ray

**Comment by dbuechel** [comment]
Okay, so you're attempting to load a configuration resource which is (I assume) behind an authentication barrier (on Moodle, I guess):

```
2021-07-08 17:21:41.617 [07] - DEBUG: [NetworkResourceLoader] Sending GET request for 'https://cybersec.scem.westernsydney.edu.au/mod/quiz/accessrule/seb/config.php?cmid=2791'...
2021-07-08 17:21:41.744 [07] - DEBUG: [NetworkResourceLoader] Received response '200 - OK'.
2021-07-08 17:21:41.746 [07] - DEBUG: [NetworkResourceLoader] The  response data is HTML.
2021-07-08 17:21:41.747 [07] - INFO: [ConfigurationRepository] Tried to load data from 'sebs://cybersec.scem.westernsydney.edu.au/mod/quiz/accessrule/seb/config.php?cmid=2791' using NetworkResourceLoader -> Result: LoadWithBrowser.
2021-07-08 17:21:41.749 [07] - INFO: The configuration resource needs authentication or is a webpage, using 'sebs://cybersec.scem.westernsydney.edu.au/mod/quiz/accessrule/seb/config.php?cmid=2791' as start URL for the browser.
```

Thus the default settings are used until authentication, successful download and reconfiguration with the resource referenced by the URL above. As the default settings prohibit the usage of virtual machines, the validation fails and SEB cannot be started:

```
2021-07-08 17:21:41.761 [07] - INFO: Validating virtual machine policy...
2021-07-08 17:21:41.765 [07] - DEBUG: [VirtualMachineDetector] Computer 'CYLABS-DEV-3' appears to be a virtual machine.
2021-07-08 17:21:41.767 [07] - ERROR: Detected virtual machine while SEB is not allowed to be run in a virtual machine! Aborting...
```

**Comment by UncleRaymondo** [comment]
Yep, it's actually launched from Moodle from a Quiz. The student clicks on Launch Safe Exam Browser.

How do I change the default settings to get around this? Would it be to load a default config on the machine inside %AppData%\SafeExamBrowser?

**Comment by UncleRaymondo** [comment]
Ok I tried launching Safe Exam Browser directly on the machine, it failed with the VM Error. Opened the SEB Configuration Tool, saved a default config allowing VM Access. Launched Safe Exam Browser again directly from the machine and it worked in that it launched SEB and came up saying it's installed but not configured, which is expected.

Trying to launch SEB from the Moodle quiz again now after the default setting is saved, and the same error appears noting VM not allowed.

The respective log files are below

[2021-07-08_18h14m46s_Client.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6782855/2021-07-08_18h14m46s_Client.log)
[2021-07-08_18h14m46s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6782856/2021-07-08_18h14m46s_Runtime.log)
[2021-07-08_18h15m21s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/6782857/2021-07-08_18h15m21s_Runtime.log)

**Comment by dbuechel** [comment]
I think there are two options:

1. Distribute a configuration file which allows VMs and leads to the quiz in Moodle. The students then log into Moodle and reconfigure SEB with the actual exam configuration stored within Moodle.
2. Place a local client configuration directly on each VM which allows the usage of virtual machines and leads to the quiz in Moodle.

**Comment by dbuechel** [comment]
I can't exactly follow what happened with the second runtime log you posted above, but again, if the flag is set correctly in a configuration, then the entire check will be skipped. So it is very unlikely that this is an issue within SEB but seems rather to be an issue with handling the configuration(s) correctly.

**Comment by UncleRaymondo** [comment]
It's a VDI Infrastructure so the students use VMWare Horizon Client to access the labs. If I launch the SEB utility now directly from the machine it works fine and starts up.

I've configured Moodle to now just use the SEB Client Config with a Browser Key - just need to tell the students that come quiz or exam time, they need to launch SEB and browse to the quiz that way... I can't see any other way or why it's not handling the configuration properly

**Comment by dbuechel** [comment]
It can get a bit complicated once there are multiple configurations involved via reconfiguration of SEB. Maybe drawing a flowchart with your specific use case and all the different configurations involved might help.

I'll convert the issue into a discussion so that you may share any further insights or questions.
